sys/mips/include/cheriasm.h

```diff
@@ -395,4 +395,57 @@
 	nop; /* Fill branch delay slot for old harware*/	\
 	.set pop;
 
+/* Derive the initial DDC/KDC and PCC/KCC from an omnipotent boot
+ * capability, presumed to be in DDC.  Clobbers C27 and C28, which surely
+ * want to be cleared before handing off control.  (We don't do so here
+ * because in locore.S we have more privileged caps to derive, and we'll
+ * clean up later.)
+ *
+ * TODO: The capabilities derived for DDC/KDC and for PCC/KCC cover the
+ * entirety of the kernel.  Acutally changing base/length requires changes
+ * in linkage.
+ */
+#define CHERI_LOCORE_ROOT_CAPS \
+	/* Grab the initial omnipotent capability */                 \
+	cgetdefault	CHERI_REG_C28;                               \
+	                                                             \
+	/* Create a reduced DDC.  */                                 \
+	cmove		CHERI_REG_C27, CHERI_REG_C28;                \
+	REG_LI		t0, CHERI_CAP_KERN_BASE;                     \
+	csetoffset	CHERI_REG_C27, CHERI_REG_C27, t0;            \
+	REG_LI		t0, CHERI_CAP_KERN_LENGTH;                   \
+	csetbounds	CHERI_REG_C27, CHERI_REG_C27, t0;            \
+	REG_LI		t0, CHERI_PERMS_KERNEL_DATA;                 \
+	candperm	CHERI_REG_C27, CHERI_REG_C27, t0;            \
+	                                                             \
+	/* Preserve a copy in KDC for exception handlers. */         \
+	csetkdc		CHERI_REG_C27;                               \
+	                                                             \
+	/* Install the new DDC. */                                   \
+	csetdefault	CHERI_REG_C27;                               \
+	                                                             \
+	/* Create a reduced PCC.  */                                 \
+	cgetpcc		CHERI_REG_C27;                               \
+	REG_LI		t0, CHERI_CAP_KERN_BASE;                     \
+	csetoffset	CHERI_REG_C27, CHERI_REG_C27, t0;            \
+	REG_LI		t0, CHERI_CAP_KERN_LENGTH;                   \
+	csetbounds	CHERI_REG_C27, CHERI_REG_C27, t0;            \
+	REG_LI		t0, CHERI_PERMS_KERNEL_CODE;                 \
+	candperm	CHERI_REG_C27, CHERI_REG_C27, t0;            \
+	                                                             \
+	/* Preserve a copy in KCC for exception handlers.  */        \
+	                                                             \
+	csetkcc		CHERI_REG_C27;                               \
+	                                                             \
+	/* Install the new PCC. */                                   \
+	REG_LI		t0, CHERI_CAP_KERN_BASE;                     \
+	cgetpcc		CHERI_REG_C28;                               \
+	cgetoffset	t1, CHERI_REG_C28;                   /* 1 */ \
+	PTR_SUBU	t1, t1, t0;                          /* 2 */ \
+	PTR_ADDIU	t1, t1, (4 * 7);                     /* 3 */ \
```

Comment:
I wonder if we can use a label difference here? E.g. `.Ltarget - .Lgetpcc_inst`

Reply:
Probably a good idea, but I couldn't immediately figure out the syntax.

Reply:
I must be holding something wrong.  What I want to write is something like this:
```
	/* Install the new PCC. */                                   \
	REG_LI		t0, CHERI_CAP_KERN_BASE;                     \
locore_cgetpcc: \
	cgetpcc		CHERI_REG_C28;                               \
	cgetoffset	t1, CHERI_REG_C28;                   /* 1 */ \
	PTR_SUBU	t1, t1, t0;                          /* 2 */ \
	PTR_ADDIU	t1, t1, (locore_landing - locore_cgetpcc);                     /* 3 */ \
	csetoffset	CHERI_REG_C27, CHERI_REG_C27, t1;    /* 4 */ \
	cjr		CHERI_REG_C27;                       /* 5 */ \
	nop;                                                 /* 6 */ \
locore_landing: \
	/* 7 (land here) */
```
but this is assembling to
```
ffffffff80100410: 48 0d e1 bf           cgetoffset      $13, $c28
ffffffff80100414: 01 ac 68 2f           dsubu   $13, $13, $12
ffffffff80100418: 80 10 04 0c           lb      $16, 1036($zero)
ffffffff8010041c: 48 1b db 4f           csetoffset      $c27, $c27, $13
ffffffff80100420: 48 1b 1f ff           cjr     $c27
ffffffff80100424: 00 00 00 00           nop
```
rather than the intended
```
ffffffff80100410: 48 0d e1 bf           cgetoffset      $13, $c28
ffffffff80100414: 01 ac 68 2f           dsubu   $13, $13, $12
ffffffff80100418: 65 ad 00 1c           daddiu  $13, $13, 28
ffffffff8010041c: 48 1b db 4f           csetoffset      $c27, $c27, $13
ffffffff80100420: 48 1b 1f ff           cjr     $c27
ffffffff80100424: 00 00 00 00           nop
```

How on earth is the assembler getting a `lb` instruction instead of a `daddiu`?

Reply:
That is very odd. Does prefixing the symbols with `.L` work? If not, I would just leave the hardcoded constants and commit this fix.

Reply:
No, that generates the wonderfully helpful errors
```
<unknown>:0: error: Undefined temporary symbol
<unknown>:0: error: Undefined temporary symbol
```